rotors_gazebo_plugins/src/gazebo_magnetometer_plugin.cpp

```diff
@@ -0,0 +1,137 @@
+  // Get the pointer to the link
+  link_ = model_->GetLink(link_name);
+  if (link_ == NULL)
+    gzthrow("[gazebo_magnetometer_plugin] Couldn't find specified link \"" << link_name << "\".");
+
+  frame_id_ = link_name;
+
+  SdfVector3 ref_mag_field;
+  SdfVector3 noise_normal;
+  SdfVector3 noise_uniform_initial_bias;
+  const SdfVector3 zeros3(0.0, 0.0, 0.0);
+
+  // Retrieve the rest of the SDF parameters
+  getSdfParam<std::string>(_sdf, "magnetometerTopic", magnetometer_topic_,
+                           mav_msgs::default_topics::MAGNETIC_FIELD);
+  getSdfParam<SdfVector3>(_sdf, "refMagField", ref_mag_field, kDefaultRefMagField);
+  getSdfParam<SdfVector3>(_sdf, "noiseNormal", noise_normal, zeros3);
+  getSdfParam<SdfVector3>(_sdf, "noiseUniformInitialBias", noise_uniform_initial_bias, zeros3);
+
+  // Listen to the update event. This event is broadcast every simulation iteration.
+  this->updateConnection_ =
+      event::Events::ConnectWorldUpdateBegin(
+          boost::bind(&GazeboMagnetometerPlugin::OnUpdate, this, _1));
+
+  magnetometer_pub_ = node_handle_->advertise<sensor_msgs::MagneticField>(magnetometer_topic_, 1);
+
+  // Create the normal noise distributions
+  noise_n_[0] = NormalDistribution(0, noise_normal.X());
+  noise_n_[1] = NormalDistribution(0, noise_normal.Y());
+  noise_n_[2] = NormalDistribution(0, noise_normal.Z());
+
+  // Create the uniform noise distribution for initial bias
+  UniformDistribution initial_bias[3];
+  initial_bias[0] = UniformDistribution(-noise_uniform_initial_bias.X(),
+                                        noise_uniform_initial_bias.X());
+  initial_bias[1] = UniformDistribution(-noise_uniform_initial_bias.Y(),
+                                        noise_uniform_initial_bias.Y());
+  initial_bias[2] = UniformDistribution(-noise_uniform_initial_bias.Z(),
+                                        noise_uniform_initial_bias.Z());
+
+  // Initialize the reference magnetic field vector in world frame, taking into
+  // account the initial bias
+  mag_W_ = math::Vector3(ref_mag_field.X() + initial_bias[0](random_generator_),
+                         ref_mag_field.Y() + initial_bias[1](random_generator_),
+                         ref_mag_field.Z() + initial_bias[2](random_generator_));
+
+  // Fill the magnetometer message
+  mag_message_.header.frame_id = frame_id_;
+  mag_message_.magnetic_field.x = mag_W_.x;
+  mag_message_.magnetic_field.y = mag_W_.y;
+  mag_message_.magnetic_field.z = mag_W_.z;
+  mag_message_.magnetic_field_covariance[0] = noise_normal.X() * noise_normal.X();
+  mag_message_.magnetic_field_covariance[4] = noise_normal.Y() * noise_normal.Y();
+  mag_message_.magnetic_field_covariance[8] = noise_normal.Z() * noise_normal.Z();
+}
+
+void GazeboMagnetometerPlugin::OnUpdate(const common::UpdateInfo& _info) {
+  // Get the current pose and time from Gazebo
+  math::Pose world_pose = link_->GetWorldPose();
```

Comment:
`math::Pose T_W_B =`
We use the convention here: https://github.com/ethz-asl/minkindr/wiki/Common-Transformation-Conventions